src/Host/Client/Impl/Host/RHostConnector.cs

```diff
@@ -0,0 +1,117 @@
+
+namespace Microsoft.R.Host.Client.Host {
+    internal abstract class RHostConnector : IRHostConnector {
+        private static readonly TimeSpan HeartbeatTimeout =
+#if DEBUG
+            // In debug mode, increase the timeout significantly, so that when the host is paused in debugger,
+            // the client won't immediately timeout and disconnect.
+            TimeSpan.FromMinutes(10);
+#else
+            TimeSpan.FromSeconds(5);
+#endif
+
+        private readonly LinesLog _log;
+        private HttpClient _broker;
+        private string _interpreterId;
+
+        protected HttpClient Broker => _broker;
+
+        public bool IsDisposed { get; private set; }
+
+        protected RHostConnector(string interpreterId) {
+            _interpreterId = interpreterId;
+            _log = new LinesLog(FileLogWriter.InTempFolder("Microsoft.R.Host.BrokerConnector"));
+        }
+
+        protected void CreateHttpClient() {
+            _broker = new HttpClient(GetHttpClientHandler()) {
+                Timeout = TimeSpan.FromSeconds(30)
+            };
+
+            _broker.DefaultRequestHeaders.Accept.Clear();
+            _broker.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+        }
+
+        protected abstract HttpClientHandler GetHttpClientHandler();
+
+        protected abstract void ConfigureWebSocketRequest(HttpWebRequest request);
+
+        public virtual void Dispose() {
+            IsDisposed = true;
+        }
+
+        protected abstract Task ConnectToBrokerAsync();
+
+        public async Task PingAsync() {
+            (await _broker.PostAsync("/ping", new StringContent(""))).EnsureSuccessStatusCode();
+        }
+
+        private async Task PingWorker() {
+            try {
+                while (true) {
+                    await PingAsync();
+                    await Task.Delay(1000);
+                }
+            } catch (OperationCanceledException) {
+            } catch (HttpRequestException) {
+            }
+        }
+
+        public async Task<RHost> Connect(string name, IRCallbacks callbacks, string rCommandLineArguments = null, int timeout = 3000, CancellationToken cancellationToken = new CancellationToken()) {
+            if (IsDisposed) {
+                throw new ObjectDisposedException(typeof(LocalRHostConnector).FullName);
+            }
+
+            await TaskUtilities.SwitchToBackgroundThread();
+
+            await ConnectToBrokerAsync();
+
+            rCommandLineArguments = rCommandLineArguments ?? string.Empty;
+
+            var request = new SessionCreateRequest {
+                InterpreterId = _interpreterId,
+                CommandLineArguments = rCommandLineArguments,
+            };
+            var requestContent = new StringContent(JsonConvert.SerializeObject(request), Encoding.UTF8, "application/json");
+
+            (await _broker.PutAsync($"/sessions/{name}", requestContent, cancellationToken)).EnsureSuccessStatusCode();
```

Comment:
I was actually thinking of creating a class that wraps HttpClient, and exposes various endpoints as strongly typed methods, basically mirroring similar methods in the controller on the other side. 

What is the most appropriate exception type for wrapping - RHostDisconnectedException?

---
In reply to: [74971357](https://github.com/Microsoft/RTVS/pull/2173#discussion_r74971357) [](ancestors = 74971357)